### DISCUSSION TITLE
Keycloak javax Dependencies with Spring Boot 3

### DISCUSSION THREAD

**Discussion by cocopudding** [keycloak/keycloak#17002] [Q&A]
We are trying to upgrade our Spring Boot to the newest 3.0.2 version, and it now uses Jakarta EE. All the imports use the jakarta packages rather than javax.

For Keycloak, we are using keycloak-spring-boot-starter and keycloak-admin-client using the latest 20.0.3 version. However I believe they are still using the old javax.servlet-api rather than full Jakarta EE support. It causes some dependency issues with the latest Spring Boot.

We managed to start our application after adding these compileOnly dependencies for our Spring Boot 3 project:
compileOnly 'javax.servlet:javax.servlet-api:4.0.1'
compileOnly 'com.sun.activation:jakarta.activation:2.0.1'

However, now from our HttpServletRequest, there is no KeycloakSecurityContext attribute, and userPrincipal is also null.

Are there new ways to using Keycloak clients with Spring Boot 3? Or if it is simply not supported, is there a timeline for Keycloak clients to have Jakarta EE support?

Thanks!

**Comment by vramik** [comment]
Just FYI: there is a plan to support Jakarta EE with upgrade to Quarkus 3. There is wip branch with that work: https://github.com/keycloak/keycloak/tree/quarkus3

**Comment by kkcmadhu** [comment]
when is it expected will it be in keycloak 22 or will this be keycoak 21.x?
will the client libraries be changed as well?

**Reply by vramik** [reply]
Currently the aim is keycloak 22, but nothing is set into a stone.

**Reply by kkcmadhu** [reply]
thanks @vramik  ,any expected time lines for keycloak 22 (of course nothing is set  into stone!).

**Reply by hantsy** [reply]
Mostly we use Keycloak as an **OAuth2 authorization server**, and your API/services serve as an **OAuth2 resource server**, in Spring Boot, no need to add any Keycloak apis. 

Check my notes: https://github.com/hantsy/spring-security-oauth2-resourceserver-example/blob/master/docs/keycloak.md

If your application is a general web application as an **OAuth2 Client**, integrate Keycloak, check the configuration here, https://github.com/hantsy/spring-security-oauth2-client-example/blob/master/client/src/main/resources/application.properties#L30C1-L40C87 no need any Keycloak API.

I also used Keycloak Admin API to do some operations on users in my project, worked well. I am using the latest Keycloak 25.0.1.

**Reply by kKapalka** [reply]
@hantsy Thanks for letting me look into the code! However, our situation is kinda more complex, we're letting our clients dynamically create keycloak realms and accomodate their respective login processes, single sign-on among them. So, while helpful, the resources on integrating Keycloak into applications using Spring Boot 3 don't really cut ut

**Reply by hantsy** [reply]
> we're letting our clients dynamically create keycloak realms and accomodate their respective login processes, single sign-on among them

Most of the API operations can be done by `org.keycloak:keycloak-admin-client`, we mainly use it to manage users in our project.
